client/shaders/second_stage/opengl_fragment.glsl

```diff
@@ -80,7 +82,15 @@ vec3 applySaturation(vec3 color, float factor)
 void main(void)
 {
 	vec2 uv = varTexCoord.st;
+#ifdef ENABLE_SSAA	
+	vec4 color = 0.25 * (
+			texture2D(rendered, uv + vec2(-1., -1.) * 0.25 * texelSize0).rgba +
+			texture2D(rendered, uv + vec2(1., -1.) * 0.25 * texelSize0).rgba +
+			texture2D(rendered, uv + vec2(-1., 1.) * 0.25 * texelSize0).rgba +
+			texture2D(rendered, uv + vec2(1., 1.) * 0.25 * texelSize0).rgba);
```

Comment:
I'm probably missing something, but for simple 2x SSAA implementation I'd expect something like:
```glsl
	vec4 color = 0.25 * (
			texture2D(rendered, uv + vec2(0.0, 0.0) * texelSize0).rgba +
			texture2D(rendered, uv + vec2(0.5, 0.0) * texelSize0).rgba +
			texture2D(rendered, uv + vec2(0.5, 0.5) * texelSize0).rgba +
			texture2D(rendered, uv + vec2(0.0, 0.5) * texelSize0).rgba);
```
(Maybe with half-pixel offset.)
Is there a reason you've decided to go with this "plus" shape?

Reply:
There is no “plus” shape. The samples are in square corners, spaced by `0.5*texelSize0`, but centered at `uv` unlike your sample. Assuming `texelSize0` is the size of *destination* pixel, the code is correct.

OTOH it could leverage built-in linear filtering and take only 2 samples, at ±{0, 0.25} (or ±{0.25, 0}) of `texelSize0`. Not sure does Irrlicht support that currently, I think I saw some code that disables filtration for RTT textures. Or whether it makes any sense at all with actual hardware.

Or it could do `texelFetch`es instead of sampling.

Reply:
Wait, I miss something. It seems the code in `secondstage.cpp` enables linear filter but this shader still does 4 samples at locations that are either exact texel centers (so the filter doesn’t add anything) or weird. @x2048 ?

Reply:
> There is no “plus” shape.

Brain fart. Somehow I thought that these offsets are like (0.25, 0.0), (0.0, 0.25), etc.

_Note to self: Don't ask questions before your morning coffee._

Reply:
@numberZero I've changed the shader to make full use of bilinear filtering. For 2x2 it will make a single sample.